common/http_test.go

```diff
@@ -0,0 +1,245 @@
+			assert.Failf(t, "expecting a successful json.Unmarshal", "#%d", i)
+			continue
+		}
+
+		assert.Equal(t, reflect.DeepEqual(recv, blankErrResponse), false, "expecting a non-blank error response")
+
+		// Now test with an error that's .HTTPCode() int conforming
+
+		// Reset w
+		w = httptest.NewRecorder()
+
+		common.WriteError(w, common.ErrorWithCode(errors.New("foo"), tt.code))
+		assert.Equal(t, w.Code, tt.code, fmt.Sprintf("#%d: got=%d want=%d", i, w.Code, tt.code))
+	}
+}
+
+type marshalFailer struct{}
+
+var errFooFailed = errors.New("foo failed here")
+
+func (mf *marshalFailer) MarshalJSON() ([]byte, error) {
+	return nil, errFooFailed
+}
+
+func TestWriteCode(t *testing.T) {
+	codes := [...]int{
+		0: http.StatusOK,
+		1: http.StatusBadRequest,
+		2: http.StatusUnauthorized,
+		3: http.StatusInternalServerError,
+	}
+
+	for i, code := range codes {
+		w := httptest.NewRecorder()
+		common.WriteCode(w, "foo", code)
+		assert.Equal(t, w.Code, code, fmt.Sprintf("#%d", i))
+
+		// Then for the failed JSON marshaling
+		w = httptest.NewRecorder()
+		common.WriteCode(w, &marshalFailer{}, code)
+		wantCode := http.StatusBadRequest
+		assert.Equal(t, w.Code, wantCode, fmt.Sprintf("#%d", i))
+		assert.Equal(t, strings.Contains(string(w.Body.Bytes()), errFooFailed.Error()), true,
+			fmt.Sprintf("#%d: expected %q in the error message", i, errFooFailed))
+	}
+}
+
+type saver struct {
+	Foo int    `json:"foo" validate:"min=10"`
+	Bar string `json:"bar"`
+}
+
+type rcloser struct {
+	closeOnce sync.Once
+	body      *bytes.Buffer
+	closeChan chan bool
+}
+
+var errAlreadyClosed = errors.New("already closed")
+
+func (rc *rcloser) Close() error {
+	var err = errAlreadyClosed
+	rc.closeOnce.Do(func() {
+		err = nil
+		rc.closeChan <- true
+		close(rc.closeChan)
+	})
+	return err
+}
+
+func (rc *rcloser) Read(b []byte) (int, error) {
+	return rc.body.Read(b)
+}
+
+var _ io.ReadCloser = (*rcloser)(nil)
+
+func makeReq(strBody string) (*http.Request, <-chan bool) {
+	closeChan := make(chan bool, 1)
+	buf := new(bytes.Buffer)
+	buf.Write([]byte(strBody))
+	req := &http.Request{
+		Header: make(http.Header),
+		Body:   &rcloser{body: buf, closeChan: closeChan},
+	}
+	return req, closeChan
+}
+
+func TestParseRequestJSON(t *testing.T) {
+	tests := [...]struct {
+		body    string
+		wantErr bool
+		useNil  bool
+	}{
+		0: {wantErr: true, body: ``},
+		1: {body: `{}`},
+		2: {body: `{"foo": 2}`}, // Not that the validate tags don't matter here since we are just parsing
+		3: {body: `{"foo": "abcd"}`, wantErr: true},
+		4: {useNil: true, wantErr: true},
+	}
+
+	for i, tt := range tests {
+		req, closeChan := makeReq(tt.body)
+		if tt.useNil {
+			req.Body = nil
+		}
+		sav := new(saver)
+		err := common.ParseRequestJSON(req, sav)
+		if tt.wantErr {
+			assert.NotEqual(t, err, nil, fmt.Sprintf("#%d: want non-nil error", i))
+			continue
+		}
+		assert.Equal(t, err, nil, fmt.Sprintf("#%d: want nil error", i))
```

Comment:
We never check the value was stored properly on success.... maybe nitpicky but at least one check on foo.

Also, simplifying the assert checks... shorter and I get the full stack trace if there is an error.
`assert.Nil(t, err, "#%d: %+v", i, err)`